### PR TITLE
[Object Spilling] Share the same S3 session for smart_open spilling.

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

Smart open is stateless, so we should pass the session directly. Otherwise, it will create a new connection with S3 whenever it spills/restores objects (which caused 100% CPU issue). 

This indicates the smart open impl is just for S3 now.

## Related issue number

https://github.com/ray-project/ray/issues/13757

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by rkooo567** [comment]
Note: I tried to use moto3 for mocking the S3 tests, but it was really tricky (especially because there are multiple processes that access the external storage, we couldn't easily mock them). I think we should instead create the storage specific tests. (For example, tests that make sure external storage interface works as expected). This will be also easier for external contributors to contribute new external storages (since they can make sure their implementation is correct without using it with Ray).

File: python/ray/external_storage.py [hunk 1]
```diff
@@ -287,6 +287,9 @@ class ExternalStorageSmartOpenImpl(ExternalStorage):
     For example, if your uri is a local file path, you should pre-create
     the directory.
 
+    Currently, this plugin only supports S3. Please feel free to contribute
+    to the support for other storage systems.
```
Comment (ericl) [inline python/ray/external_storage.py:291]: Why not just special case it for S3?

Comment (rkooo567) [reply python/ray/external_storage.py:291]: Yeah actually that sounds like a better idea. Let me fix the comment

File: python/ray/external_storage.py [hunk 2]
```diff
@@ -302,22 +302,32 @@ class ExternalStorageSmartOpenImpl(ExternalStorage):
     def __init__(self,
                  uri: str,
                  prefix: str = DEFAULT_OBJECT_PREFIX,
-                 override_transport_params: dict = None):
+                 override_transport_params: dict = None,
+                 is_for_s3: bool = True):
```
Comment (stephanie-wang) [inline python/ray/external_storage.py:306]:
```suggestion
                 use_s3: bool = True):
```

**Review by stephanie-wang — APPROVED** [review]

File: python/ray/external_storage.py [hunk 3]
```diff
@@ -302,22 +302,32 @@ class ExternalStorageSmartOpenImpl(ExternalStorage):
...
         try:
             from smart_open import open  # noqa
+            import boto3  # noqa
```
Comment (ericl) [inline python/ray/external_storage.py:309]: Move this import into the is_for_s3 block?

File: python/ray/external_storage.py (hunk 2 above)
Comment (ericl) [reply python/ray/external_storage.py:306]: Why not detect this based on the URI?

Comment (ericl) [reply python/ray/external_storage.py:306]: .startswith("s3")

Comment (rkooo567) [reply python/ray/external_storage.py]: That's a good idea. Done.

**Pull request merged**